GitCommands/Git/GitModule.cs

```diff
@@ -2126,15 +2126,11 @@ public IEnumerable<string> GetSettings(string setting)
             return LocalConfigFile.GetValues(setting);
         }
 
-        public string GetSetting(string setting)
-        {
-            return LocalConfigFile.GetValue(setting);
-        }
+        public string GetSetting(string setting) => LocalConfigFile.GetValue(setting);
+        public T? GetSetting<T>(string setting) where T : struct => LocalConfigFile.GetValue<T>(setting);
 
-        public string GetEffectiveSetting(string setting)
-        {
-            return EffectiveConfigFile.GetValue(setting);
-        }
+        public string GetEffectiveSetting(string setting) => EffectiveConfigFile.GetValue(setting);
+        public T? GetEffectiveSetting<T>(string setting) where T : struct => EffectiveConfigFile.GetValue<T>(setting);
```

Comment:
I still have issues with this API because it is confusing and contradictory in nature:
```cs
Module.GetEffectiveSetting<bool>("rebase.autosquash") --> true/false/null
```

The implementation must match the contract - if the caller explicitly asked for `bool`, the result has to match or fail.
That is:
```cs
Module.GetEffectiveSetting<bool>("rebase.autosquash") --> true/false or throw
Module.GetEffectiveSetting<bool?>("rebase.autosquash") --> true/false/null
```

Reply:
Something like https://github.com/RussKie/gitextensions/commit/9fac4e8f1ae9fa2cefff00468998c53968a4f060.

The tests would need to set/unset test settings and would look something like this:
```cs
            /*

            Not supported:

            Console.WriteLine("CommitTemplateManager?");
            _ = new GitModule("").GetEffectiveSetting2<CommitTemplateManager?>("commit.template");
            Console.WriteLine("CommitTemplateManager");
            _ = new GitModule("").GetEffectiveSetting2<CommitTemplateManager>("commit.template");

            */

            /*

              Standard Output:
                int?:
                bool?:
                string: false
                bool: False

             */
            Console.WriteLine($"int?: {new GitModule("").GetEffectiveSetting<int?>("commit.template")}");
            Console.WriteLine($"bool?: {new GitModule("").GetEffectiveSetting<bool?>("commit.template")}");

            Console.WriteLine($"string: {new GitModule("").GetEffectiveSetting<string>("rebase.updateRefs")}");
            Console.WriteLine($"bool: {new GitModule("").GetEffectiveSetting<bool>("rebase.updateRefs")}");

            /*

            System.FormatException : Input string was not in a correct format:

            Console.WriteLine($"int: {new GitModule("").GetEffectiveSetting2<int>("rebase.updateRefs")}");

            */
```

Reply:
(Reminder) Current api designed:

```csharp
Module.GetEffectiveSetting<bool>("rebase.autosquash") --> true/false/null (not set) or throw (conversion failed => bug) 
```

I understand your point of view (and in other use cases I would agree) but I think the current api is delivering much value and is easier to work with.

Note: for me you should see T more as the type of data your setting is storing so it make sense to receive something null when the setting is not set.

But the added value compare to your proposition is that most of the checks are made at compile time.
T must be a value type so it is not building whereas in your proposition, you will have the error at runtime.

But the current api oblige you to handle the `null` case (i.e. no settings).
In your proposition, use case like `GetEffectiveSetting<bool>()` where your expect a bool will increase a lot the probability to not manage the case where there is no settings and will crash in production for someone else whereas it was working well during your developer testing because the dev has the setting and not the user.

I don't see a case where you will pass something not nullable except if you want to shoot yourself into the foot.

So, if you forgot the little discrepancy between the type specified and the one returned (but once again, if we think about it from the git setting point of view, it makes sense), I think that working with the current api **is easier because more bug proof** than the proposed api.

Also, I find the your solution add a lot of not straight forward code not easy to understand.

Reply:
Fair enough